cmd/issueZKCert.go

```diff
@@ -310,3 +309,81 @@ func encodeMerkleProof(proof merkle.Proof) [][32]byte {
 
 	return res
 }
+
+func registerToQueue(
+	ctx context.Context,
+	client *ethclient.Client,
+	providerKey *ecdsa.PrivateKey,
+	recordRegistry RecordRegistryTransactor,
+	leafHash zkcertificate.Hash,
+) (*types.Transaction, error) {
+	chainID, err := client.ChainID(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("retrieve chain id: %w", err)
+	}
+
+	auth, err := bind.NewKeyedTransactorWithChainID(providerKey, chainID)
+	if err != nil {
+		return nil, fmt.Errorf("create transaction signer from private key: %w", err)
+	}
+
+	tx, err := recordRegistry.RegisterToQueue(auth, leafHash.Bytes32())
+	if err != nil {
+		exists, checkErr := recordRegistry.CheckZkCertificateHashInQueue(&bind.CallOpts{}, leafHash.Bytes32())
+		if checkErr != nil {
+			return nil, fmt.Errorf("register to queue failed: %w, also failed to check if zkCertificateHash is in queue: %w", err, checkErr)
+		}
+		if exists {
+			return nil, nil
+		}
+		return nil, fmt.Errorf("register to queue failed: %w", err)
+	}
+
+	return tx, nil
+}
+
+func registerAndWaitForZkCertificateTurn(
+	ctx context.Context,
+	client *ethclient.Client,
+	providerKey *ecdsa.PrivateKey,
+	registry RecordRegistry,
+	leafHash zkcertificate.Hash,
+) error {
+	registerTx, err := registerToQueue(ctx, client, providerKey, registry, leafHash)
+	if err != nil {
+		return fmt.Errorf("register zkCertificate hash to queue: %w", err)
+	}
+
+	if registerTx != nil {
+		receipt, err := bind.WaitMined(ctx, client, registerTx)
+		if err != nil {
+			return fmt.Errorf("wait until queue registration transaction is mined: %w", err)
+		}
+		if receipt.Status == 0 {
+			return fmt.Errorf("queue registration transaction %q failed", receipt.TxHash)
+		}
+	}
+
+	for {
+		startTime, expirationTime, err := registry.GetTimeParameters(&bind.CallOpts{}, leafHash.Bytes32())
+		if err != nil {
+			return fmt.Errorf("retrieve time parameters for zkCertificate hash: %w", err)
+		}
+
+		currentTime := big.NewInt(time.Now().Unix())
+		if startTime.Cmp(currentTime) <= 0 {
+			break
+		}
+		if expirationTime.Cmp(currentTime) <= 0 {
+			return fmt.Errorf("queue wait time expired for zkCertificate hash %s", leafHash)
+		}
+
+		select {
+		case <-time.After(5 * time.Second):
```

Comment:
Maybe we should wait until `min(startTime, now() + 5 seconds)`?

Reply:
It is safe as expiration time is a few minutes.